### DISCUSSION TITLE
Add an option like `noImplicitAny` in TypeScript compiler

### DISCUSSION THREAD

**Discussion by mmohaveri** [microsoft/pyright#5589] [Ideas]
In TypeScript compiler has a `noImplicitAny` option which raises if a variable has the `Any` type without explicitly being marked as one.

This will ensure that all of your code is type safe unless you explicitly decide to circumvent type checking by using the Any type.
As far as I can see, there's no such configuration in pyright.

This can be useful if libraries you use do not have a good enough type hints and they leak `Any` into your codebase.
For example, in SQLAlchemy, a row (in row.Row) defines a getter (`__getattr__`) without any type annotation, so the result of `row.SomeColumnName` will be `Any`.

Not this practically disables (or reduces) type checking for all following expressions that use this value, it also introduces a possible runtime error an `row.SomeNotExistingColumn` will not cause any type error, but raises in runtime.

If I'm able to disallow any implicit `Any` in my codebase, I'll be forced to either explicitly accept these risks by marking the variable that holds the result of `row.SomeColumnName` as 'Any`, i.e: `some_var: Any = row.SomeColumnName`, or use another API of the SQLAlchemy (like row.tuple()) which has type annotations.

**Comment by erictraut** [comment]
Pyright already has this support. An "implicit Any" is referred to as `Unknown`, and pyright will report uses of `Unknown` in various situations based on which of the `reportUnknownXXX` diagnostic checks you've enabled. In "strict" type checking mode, all of these are enabled.

If a library uses an explicit `Any` annotation, there's not much that pyright can do about that. `Any` is a supported part of the python type system, and type checkers need to honor that.

**Comment by mmohaveri** [comment]
I understand that type checker needs to honer the library's `Any`, but I think it would be nice to have an option to disallow leakage of these `Any`s into my code base without explicitly allowing it in code by annotating the variable as `Any`.

**Reply by erictraut** [reply]
Thanks for your suggestion. A potential way of implementing this would be to have an option that causes all `Any` type from a library to be treated as `Unknown`.

I'm not going to implement this feature now, but I might consider it in the future if there is sufficient demand (e.g. in the form of upvotes) from pyright users.
